package NYT.NChunkClient.NProto;

import "yt/core/misc/guid.proto";
import "yt/core/misc/protobuf_helpers.proto";
import "yt/ytlib/chunk_client/chunk_spec.proto";

////////////////////////////////////////////////////////////////////////////////

message TBlockInfo
{
    required int64 offset = 1;
    required int32 size = 2;
    required uint64 checksum = 3;
}

// Chunk meta extension.
message TBlocksExt
{
    repeated TBlockInfo blocks = 1;
}

// Used within sliced input chunks to override sizes provided in TMiscExt.
message TSizeOverrideExt
{
    required int64 row_count = 1;
    required int64 uncompressed_data_size = 2;
}

// Chunk meta extension.
message TMiscExt
{
    // Total size of uncompressed blocks.
    optional int64 uncompressed_data_size = 1 [default = 0];

    // Total size of blocks with data sent to holder.
    optional int64 compressed_data_size = 5 [default = 0];

    // Total size of (key column names + values + 1) in each row.
    // Must be greater than zero.
    optional int64 data_weight = 7 [default = 1];

    // Size of chunk meta (without TMiscExt), obtained via ByteSize.
    optional int64 meta_size = 6 [default = 0];

    // Number of rows in this chunk (tables only).
    optional int64 row_count = 2 [default = 0];

    optional int32 compression_codec = 3 [default = 0];

    // Indicates if the chunk contains a sorted rowset (tables only).
    optional bool sorted = 4 [default = false];

    // Number of key-value pairs across all rows in the chunk.
    optional int64 value_count = 8 [default = 0];

    // Uncompressed size of the largest block.
    optional int64 max_block_size = 9 [default = 0];
}

message TChunkInfo
{
    // Exact full size (data size plus meta size).
    // For erasure chunks this is the sum for all parts.
    required int64 disk_space = 1;

    // Not sent during full heartbeats.
    optional uint64 meta_checksum = 2;
}

message TPartInfo
{
    // Number of first block in whole chunk
    required int32 start = 1;

    // Sizes of blocks. Needed for repair.
    repeated int64 block_sizes = 2;
}

message TErasurePlacementExt
{
    repeated TPartInfo part_infos = 1;
    required int32 parity_part_count = 2;
    required int32 parity_block_count = 3;
    required int64 parity_block_size = 4;
    required int64 parity_last_block_size = 5;
}

// TChunkMeta is stored in *.meta files on data nodes
// and passed around (possibly with a different extension subset).
message TChunkMeta
{
    required int32 type = 1;
    required NYT.NProto.TExtensionSet extensions = 2;

    // Format version validated by reader.
    required int32 version = 3;
}

////////////////////////////////////////////////////////////////////////////////
